// ngx-soap/esm2015/lib/soap/soap.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/*
 * Copyright (c) 2011 Vinay Pulim <vinay@milewise.com>
 * MIT Licensed
 */
import * as tslib_1 from "tslib";
import * as wsdl from './wsdl';
import { security } from './security';
import { Client } from './client';
export { Client } from './client';
export { security } from './security';
export const /** @type {?} */ WSDL = wsdl.WSDL;
/**
 * @return {?}
 */
function createCache() {
    const /** @type {?} */ cache = {};
    return function (key, load, callback) {
        if (!cache[key]) {
            load(function (err, result) {
                if (err) {
                    return callback(err);
                }
                cache[key] = result;
                callback(null, result);
            });
        }
        else {
            // process.nextTick(function () {
            //   callback(null, cache[key]);
            // });
            Promise.resolve(true).then(() => callback(null, cache[key]));
        }
    };
}
const /** @type {?} */ getFromCache = createCache();
/**
 * @param {?} url
 * @param {?} options
 * @return {?}
 */
function _requestWSDL(url, options) {
    // if (typeof options === 'function') {
    //   callback = options;
    //   options = {};
    // }
    return wsdl.open_wsdl(url, options);
    // const openWsdl = wsdl.open_wsdl.bind(null, url, options);
    // if (options.disableCache === true) {
    //   openWsdl(callback);
    // } else {
    //   getFromCache(url, openWsdl, callback);
    // }
}
/**
 * @param {?} url
 * @param {?} options
 * @param {?} endpoint
 * @return {?}
 */
export function createClient(url, options, endpoint) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        // if (typeof options === 'function') {
        //   endpoint = callback;
        //   callback = options;
        //   options = {};
        // }
        if (typeof options === 'undefined') {
            options = {};
        }
        console.log("createClient", options);
        endpoint = options.endpoint || endpoint;
        // _requestWSDL(url, options, function(err, wsdl) {
        //   callback(err, wsdl && new Client(wsdl, endpoint, options));
        // });
        const /** @type {?} */ wsdl = yield _requestWSDL(url, options);
        const /** @type {?} */ client = new Client(wsdl, endpoint, options);
        return client;
    });
}
// export function createClientAsync(url, options, endpoint): Observable<any> {
//   if (typeof options === 'undefined') {
//     options = {};
//   }
//   console.log("createClientAsync", options)
//   return createClient(url, options, endpoint);
// }
// export function listen(server, pathOrOptions, services, xml) {
//   let options: any = {},
//     path = pathOrOptions,
//     uri = '';
//   if (typeof pathOrOptions === 'object') {
//     options = pathOrOptions;
//     path = options.path;
//     services = options.services;
//     xml = options.xml;
//     uri = options.uri;
//   }
//   const wsdl = new WSDL(xml || services, uri, options);
//   return new Server(server, path, services, wsdl, options);
// }
export const /** @type {?} */ BasicAuthSecurity = security.BasicAuthSecurity;
export const /** @type {?} */ NTLMSecurity = security.NTLMSecurity;
export const /** @type {?} */ WSSecurity = security.WSSecurity;
export const /** @type {?} */ WSSecurityCert = security.WSSecurityCert;
export const /** @type {?} */ ClientSSLSecurity = security.ClientSSLSecurity;
export const /** @type {?} */ ClientSSLSecurityPFX = security.ClientSSLSecurityPFX;
export const /** @type {?} */ BearerSecurity = security.BearerSecurity;
